### DISCUSSION TITLE
Access variable in `log:collectAllIn` pattern in rule

### DISCUSSION THREAD

**Discussion by lolgab** [eyereasoner/eye#89] [General]
I'm trying out notation3 with eyereasoner and I have a problem I don't know how to solve.
Given data for "transactions" I want to calculate the total amount of a `?thing`.
I managed to calculate the total using `log:collectAllIn` to first create a list of integers and then pass this list to `?list math:sum ?result` to sum its values.
My problem is to keep the reference to a node that I find in the pattern passed as the second value of the list that is used together with `log:collectAllIn`.
In the following example, `?thing` in the result is not the same as `?thing` in the pattern.

```turtle
@prefix log: <http://www.w3.org/2000/10/swap/log#> . 
@prefix math: <http://www.w3.org/2000/10/swap/math#> .
@prefix : <http://example.org/> .

{
  (
    ?amount
    {
      ?a
        :destination ?thing ;
        :amount ?amount .
    }
    ?amounts
  ) log:collectAllIn _:t .
  ?amounts math:sum ?amountsTotal .
} => {
  ?thing :total ?amountsTotal .
} .
```

How would you solve such problem with n3?

Thank you,
Lorenzo

**Comment by josd** [comment]
You have to bind `?thing` outside `log:collectAllIn` for instance as
```
{
  ?x :destination ?thing. 
  (
    ?amount
    {
      ?a
        :destination ?thing ;
        :amount ?amount .
    }
    ?amounts
  ) log:collectAllIn _:t .
  ?amounts math:sum ?amountsTotal .
} => {
  ?thing :total ?amountsTotal .
} .
```
and that should give `:total` triples for all different `?thing`'s

**Reply by lolgab** [reply]
Thank you very much for your help! Binding the `?thing` variable outside works indeed! It's not very clear to me why it does but I'll keep it in mind.
By the way, n3 and the eye reasoner are amazing! I'm amazed by how elegantly you can solve problems!

**Reply by lolgab** [reply]
I just realized that you can bind the `?thing` variable in whatever way you want, so, since my rule applies to nodes of a specific type I can just bind it with:

```turtle
?thing a :ThingType
```

Which makes the rule more readable.
